web/postgres_typemap.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+func toSpannerTypePostgres(srcType string, spType string, mods []int64) (ddl.Type, []internal.SchemaIssue) {
+
+	switch srcType {
+	case "bool", "boolean":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		case ddl.Int64:
+			return ddl.Type{Name: ddl.Int64}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Bool}, nil
+		}
+
+	case "bigserial":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Int64}, []internal.SchemaIssue{internal.Serial}
+		}
+	case "bpchar", "character": // Note: Postgres internal name for char is bpchar (aka blank padded char).
+		switch spType {
+		case ddl.Bytes:
+			if len(mods) > 0 {
+				return ddl.Type{Name: ddl.Bytes, Len: mods[0]}, nil
+			}
+			return ddl.Type{Name: ddl.Bytes, Len: 1}, nil
+		default:
+			if len(mods) > 0 {
+				return ddl.Type{Name: ddl.String, Len: mods[0]}, nil
+			}
+			// Note: bpchar without length specifier is equivalent to bpchar(1)
+			return ddl.Type{Name: ddl.String, Len: 1}, nil
+		}
+	case "bytea":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, nil
+		default:
+			return ddl.Type{Name: ddl.Bytes, Len: ddl.MaxLength}, nil
+		}
+	case "date":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Date}, nil
+		}
+	case "float8", "double precision":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Float64}, nil
+		}
+	case "float4", "real":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Float64}, []internal.SchemaIssue{internal.Widened}
+		}
+	case "int8", "bigint":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Int64}, nil
+		}
+	case "int4", "integer":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Int64}, []internal.SchemaIssue{internal.Widened}
+		}
+	case "int2", "smallint":
+		switch spType {
+		case ddl.String:
+			return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Widened}
+		default:
+			return ddl.Type{Name: ddl.Int64}, []internal.SchemaIssue{internal.Widened}
+		}
+	case "numeric": // Map all numeric types to float64.
```

Comment:
Let's add support for Spanner NUMERIC as part of this PR. Separately, I'll make changes to toddl.go to support NUMERIC in command line mode. In this PR, we'll just address support for NUMERIC in webui mode.

First, let's get rid of the comment:
// Map all numeric types to float64.

Second, let's change the default code to:

                          // TODO: check mod[0] and mod[1] and generate a warning
                          // if this numeric won't fit in Spanner's NUMERIC.
			return ddl.Type{Name: ddl.Numeric}, nil

I'll deal with this TODO later -- this is already a huge improvement over mapping to FLOAT64!

Reply:
sure Nevin, i'll make this change in mysql_typemap and postgres_typemap, one question, should i map decimal to NUMERIC as well in mysql_typemap, currently both numeric and decimal are mapped to FLOAT64?

Reply:
Yes, both should map to NUMERIC. They are essentially the same (I think they map to the same implementation internally).

Reply:
For reference: [PR98](https://github.com/cloudspannerecosystem/harbourbridge/pull/98) for support of NUMERIC in command line mode.